contributors/design-proposals/pod-priority-api.md

```diff
@@ -0,0 +1,237 @@
+the key metrics that provides the information to the cluster. This document is a
+more detailed design proposal for part of the high-level architecture described
+in [Resource sharing architecture for batch and serving workloads in Kubernetes](https://docs.google.com/document/d/1-H2hnZap7gQivcSU-9j4ZrJ8wE_WwcfOkTeAGjzUyLA).
+
+## Overview 
+
+This design doc introduces the concept of priorities for pods in Kubernetes and
+how the priority impacts scheduling and preemption of pods when the cluster
+runs out of resources. A pod can specify a priority at the creation time. The
+priority must be one of the valid values and there is a total order on the
+values. The priority of a pod is independent of its workload type. The priority
+is global and not specific to a particular namespace.
+
+## Detailed Design 
+
+### Effect of priority on scheduling 
+
+One could generally expect a pod with higher priority has a higher chance of
+getting scheduled than the same pod with lower priority. However, there are
+many other parameters that affect scheduling decisions. So, a high priority pod
+may or may not be scheduled before lower priority pods. The details of
+what determines the order at which pods are scheduled are beyond the scope of
+this document.
+
+### Effect of priority on preemption 
+
+Generally, lower priority pods are more likely to get preempted by higher
+priority pods when cluster has reached a threshold. In such a case, scheduler
+may decide to preempt lower priority pods to release enough resources for higher
+priority pending pods. As mentioned before, there are many other parameters
+that affect scheduling decisions, such as affinity and anti-affinity. If
+scheduler determines that a high priority pod cannot be scheduled even if lower
+priority pods are preempted, it will not preempt lower priority pods. Scheduler
+may have other restrictions on preempting pods, for example, it may refuse to
+preempt a pod if PodDisruptionBudget is violated. The details of scheduling and
+preemption decisions are beyond the scope of this document.
+
+### Priority in PodSpec 
+
+Pods may have priority in their pod spec. PodSpec will have two new fields
+called "PriorityClassName" which is specified by user, and "Priority" which will
+be populated by Kubernetes. User-specified priority (PriorityClassName) is a 
+string and all of the valid priority classes are defined by a system wide
+mapping that maps each string to an integer. The PriorityClassName specified in
+a pod spec must be found in this map or the pod creation request will be
+rejected. If PriorityClassName is empty, it will resolve to the default
+priority (See below for more info on name resolution). Once the 
+PriorityClassName is resolved to an integer, it is placed in "Priority" field of
+PodSpec.
+
+
+```
+type PodSpec struct {
+  ...
+  PriorityClassName string
+  Priority          int32  // Populated by Admission Controller. Users are not allowed to set it directly.
```

Comment:
Priority was initially in the status, but several reviewers objected that design and I then moved it to the spec. The main concerns were: 1) a change in pod spec would result in a change to pod status, 2) a misbehaving component that can change pod status could change pod priority.
Here is what they said in their own words:

@thockin:
> There's a rule of thumb that everything in status can be recreated from observation. This would not be the case here, as (I assume) priority name->num mapping could change after a pod was created.

@liggitt 
> Currently, setting the spec explicitly does not affect the status (and vice versa) (the create/update strategy for all objects with spec preserves the existing status as-is, and the status strategy preserves the existing spec as-is). This changes that behavior for pods, and means that modifying spec can conflict with status updates, and anything that can modify status can change "effective" pod priority.

@wojtek-t 
> What @liggitt wanted to write is that is that if e.g. you decide to update PriorityName in PodSpec, admission controller will update the Priority field in Status. So an update to Spec also results in update to Status.
I think I share the same concern - I'm also not aware of any other such pattern anywhere.

@gmarek 
> In general I believe that Spec is supposed to have all the data about "desired state", and what you suggest it adding some of it to Status, which if I'm not mistaken is never used in such context. I know that the value in Status can be inferred from Spec + external data, but somehow explicitly using stuff from Status to learn about user's intent feels wrong.